crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -167,15 +168,244 @@ pub unsafe trait SystemParam: Sized {
+    #[inline]
+    #[allow(unused_variables)]
+    fn apply(state: &mut Self::State, system_meta: &SystemMeta, world: &mut World) {}
+
+    /// # Safety
+    ///
+    /// This call might use any of the [`World`] accesses that were registered in [`Self::init_state`].
+    /// You must ensure that none of those accesses conflict with any other [`SystemParam`]s running in parallel with this one.
+    unsafe fn get_param<'world, 'state>(
+        state: &'state mut Self::State,
+        system_meta: &SystemMeta,
+        world: &'world World,
+        change_tick: u32,
+    ) -> Result<
+        Self::Item<'world, 'state>,
+        <Self::Item<'world, 'state> as ResultfulSystemParam>::Error,
+    >;
+}
+
+unsafe impl<T: ResultfulSystemParam> OptionalSystemParam for T {
+    type State = T::State;
+
+    type Item<'world, 'state> = T::Item<'world, 'state>;
+
+    fn init_state(world: &mut World, system_meta: &mut SystemMeta) -> Self::State {
+        T::init_state(world, system_meta)
+    }
+
+    fn new_archetype(state: &mut Self::State, archetype: &Archetype, system_meta: &mut SystemMeta) {
+        T::new_archetype(state, archetype, system_meta);
+    }
+
+    fn apply(state: &mut Self::State, system_meta: &SystemMeta, world: &mut World) {
+        T::apply(state, system_meta, world);
+    }
+
+    unsafe fn get_param<'world, 'state>(
+        state: &'state mut Self::State,
+        system_meta: &SystemMeta,
+        world: &'world World,
+        change_tick: u32,
+    ) -> Option<Self::Item<'world, 'state>> {
+        T::get_param(state, system_meta, world, change_tick).ok()
+    }
+}
+
+unsafe impl<T: ResultfulSystemParam> SystemParam for Result<T, T::Error> {
+    type State = T::State;
+
+    type Item<'world, 'state> =
+        Result<T::Item<'world, 'state>, <T::Item<'world, 'state> as ResultfulSystemParam>::Error>;
+
+    fn init_state(world: &mut World, system_meta: &mut SystemMeta) -> Self::State {
+        T::init_state(world, system_meta)
+    }
+
+    fn new_archetype(state: &mut Self::State, archetype: &Archetype, system_meta: &mut SystemMeta) {
+        T::new_archetype(state, archetype, system_meta);
+    }
+
+    fn apply(state: &mut Self::State, system_meta: &SystemMeta, world: &mut World) {
+        T::apply(state, system_meta, world);
+    }
+
+    unsafe fn get_param<'world, 'state>(
+        state: &'state mut Self::State,
+        system_meta: &SystemMeta,
+        world: &'world World,
+        change_tick: u32,
+    ) -> Self::Item<'world, 'state> {
+        T::get_param(state, system_meta, world, change_tick)
+    }
+}
+
 /// A [`SystemParam`] that only reads a given [`World`].
 ///
 /// # Safety
 /// This must only be implemented for [`SystemParam`] impls that exclusively read the World passed in to [`SystemParam::get_param`]
 pub unsafe trait ReadOnlySystemParam: SystemParam {}
 
+unsafe impl<T: ReadOnlySystemParam + OptionalSystemParam> ReadOnlySystemParam for Option<T> {}
```

Comment:
```suggestion
// SAFETY: the system param is read-only by the trait bound
// and so maybe getting it will never cause write-access to be granted.
unsafe impl<T: ReadOnlySystemParam + OptionalSystemParam> ReadOnlySystemParam for Option<T> {}
```